Sources/SecureXPC/Codable Property Wrappers/File Descriptor XPC Wrappers.swift

```diff
@@ -0,0 +1,184 @@
+    }
+    
+    var descriptor: CInt {
+        wrappedValue
+    }
+    
+    func close() throws {
+        let closeResult = Darwin.close(wrappedValue)
+        guard closeResult == 0 else {
+            throw XPCError.unknown(description: """
+            Unable to close file descriptor: \(wrappedValue)
+            Result code: \(closeResult)
+            """)
+        }
+    }
+}
+
+// MARK: FileDescriptor
+
+/// Wraps a [`FileDescriptor`](https://developer.apple.com/documentation/system/filedescriptor) such that it can be sent over an XPC
+/// connection.
+///
+/// By default the provided file descriptor will be closed once it has been encoded.
+///
+/// When creating an ``XPCRoute`` that directly transfers this type as either the message or reply type, `FileDescriptorForXPC` must be the specified type,
+/// not `FileDescriptor`. This is not applicable when transferring a type which _contains_ a wrapped file descriptor as one of its properties.
+///
+/// This property wrapper shares an underlying representation with ``DarwinFileDescriptorForXPC`` and ``FileHandleForXPC`` and therefore may be
+/// used interchangeably between the server and client. However, to make use of such functionality requires routes with identical names and differing message and/or
+/// reply types.
+@available(macOS 11.0, *)
+@propertyWrapper public struct FileDescriptorForXPC {
+    public var wrappedValue: FileDescriptor
+    fileprivate let closeOnEncode: Bool
+    
+    public init(wrappedValue: FileDescriptor, closeOnEncode: Bool = true) {
+        self.wrappedValue = wrappedValue
+        self.closeOnEncode = closeOnEncode
+    }
+}
+
+@available(macOS 11.0, *)
+extension FileDescriptorForXPC: FileDescriptorCodable {
+    init(descriptor: CInt, closeOnEncode: Bool) {
+        self.wrappedValue = FileDescriptor(rawValue: descriptor)
+        self.closeOnEncode = closeOnEncode
+    }
+    
+    var descriptor: CInt {
+        wrappedValue.rawValue
+    }
+    
+    func close() throws {
+        try wrappedValue.close()
+    }
+}
+
+// MARK: FileHandle
+
+/// Wraps a [`FileHandle`](https://developer.apple.com/documentation/foundation/filehandle) such that it can be sent over an XPC
+/// connection.
+///
+/// By default the provided file handle will be closed once it has been encoded.
+///
+/// When creating an ``XPCRoute`` that directly transfers this type as either the message or reply type, `FileHandleForXPC` must be the specified type, not
+/// `FileHandle`. This is not applicable when transferring a type which _contains_ a wrapped file handle as one of its properties.
+///
+/// This property wrapper shares an underlying representation with ``DarwinFileDescriptorForXPC`` and ``FileDescriptorForXPC`` and therefore
+/// may be used interchangeably between the server and client. However, to make use of such functionality requires routes with identical names and differing
+/// message and/or reply types.
+@propertyWrapper public struct FileHandleForXPC {
```

Comment:
How do you feel about using `Foundation.FileHandle` as the fundamental file-descriptor abstraction, instead of a property-wrapped plain `Int32`/`CInt`?

I think we should be opinionated and only support one of the two, and of the two, I think `FileHandle` is better.

I think this has two upsides and one downside:

1. Pro: the wrapped type is more specific. It's a specific struct, rather than a general integer.
2. Pro: the type already exists, and has `closeOnDealloc` just like `cloneOnEncode`. We don't need to keep our own bool or do our own close logic.
3. Con: It makes `Foundation` a required dependency, but I don't think that matters given that we already require XPC and are necessarily on Apple platforms that have Foundation.

Furthermore: I would just suggest simply doing `extension FileHandle: FileDescriptorCodable`, and removing the complexity of a property wrapper from this API design. It's perfectly safe to do this kind of retroactive modelling, as long as you're the authority over either the conforming type (We're not, Apple own `Foundation.FileHandle`) or the protocol (we are)

Reply:
Let me think about it and will reply with more thoughts later.

Reply:
> How do you feel about using Foundation.FileHandle as the fundamental file-descriptor abstraction, instead of a property-wrapped plain Int32/CInt?

Do you mean internally or that there be no publicly exposed API for using POSIX file descriptors?

> Pro: the type already exists, and has `closeOnDealloc` just like `cloneOnEncode`. We don't need to keep our own bool or do our own close logic.

---

I'd like to support `System.FileDescriptor` as well as it doesn't appear to have a `closeOnDealloc` equivalent as far as I can tell. And it goes as far as to say:
> You are responsible for managing the lifetime and validity of FileDescriptor values, in the same way as you manage a raw C file handle.

---

> I would just suggest simply doing `extension FileHandle: FileDescriptorCodable`, and removing the complexity of a property wrapper from this API design. 

How would one do that? I tried this a while back directly with `Decodable` (which ought to be an equivalent situation), but it appeared not to be possible because it has a required initializer which can't be added via an extension:

> 'required' initializer must be declared directly in class 'FileHandle' (not in an extension)

Reply:
> Do you mean internally or that there be no publicly exposed API for using POSIX file descriptors?

no publicly exposed API for using POSIX file descriptors.

> I'd like to support System.FileDescriptor as well as it doesn't appear to have a closeOnDealloc equivalent as far as I can tell.

True, so we'll need one type for `Foundation.FileHandle` and one for `System.FileDescriptor`

> but it appeared not to be possible because it has a required initializer which can't be added via an extension

Ah... that might be a show stopper for that idea.